OpenRA.Mods.Common/AI/BaseBuilder.cs

```diff
@@ -37,14 +42,44 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
 			playerPower = pm;
 			playerResources = pr;
 			this.category = category;
+			cachedBases = 0;
+			waterAvailable = false;
+			checkedWater = false;
+			failCount = 0;
+			failRetryTicks = ai.Info.StructureProductionResumeDelay;
 		}
 
 		public void Tick()
 		{
+			// If failed to place something N consecutive times, wait M ticks until resuming building production
```

Comment:
This can be simplified to
```csharp
// If failed to place something N consecutive times, wait M ticks until resuming building production
if (failCount >= ai.Info.MaximumFailedPlacementAttempts && --failRetryTicks <= 0)
    failCount = 0;

// Only update once per second or so
```